ambry-network/src/main/java/com/github/ambry/network/http2/Http2BlockingChannel.java

```diff
@@ -68,118 +45,78 @@
-            if (streamId == null) {
-              logger.error("Http2ResponseHandler unexpected message received: " + msg);
-              return;
-            }
-            ctx.channel().attr(RESPONSE_PROMISE).getAndSet(null).setSuccess(msg.content().retainedDuplicate());
-            logger.trace("Stream response received.");
-          }
-        });
-      }
-    });
   }
 
   @Override
   public void disconnect() throws IOException {
-    channel.disconnect().syncUninterruptibly();
-    workerGroup.shutdownGracefully();
+
   }
 
+  /**
+   * A new stream channel is created for each send.
+   */
   @Override
   public void send(Send request) throws IOException {
-    ByteBufferChannel byteBufferChannel = new ByteBufferChannel(ByteBuffer.allocate((int) request.sizeInBytes()));
-    while (!request.isSendComplete()) {
-      request.writeTo(byteBufferChannel);
+    Channel streamChannel;
+    try {
+      streamChannel = http2MultiplexedChannelPool.acquire().get(500, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      throw new IOException("Can't acquire stream channel: ", e);
     }
-    byteBufferChannel.getBuffer().position(0);
-    ByteBuf byteBuf = Unpooled.wrappedBuffer(byteBufferChannel.getBuffer());
 
-    Http2StreamChannel childChannel = http2StreamChannelBootstrap.open().syncUninterruptibly().getNow();
-    Http2Headers http2Headers = new DefaultHttp2Headers().method(HttpMethod.POST.asciiName()).scheme("https").path("/");
-    responsePromise = childChannel.eventLoop().newPromise();
-    childChannel.attr(RESPONSE_PROMISE).set(responsePromise);
+    ChannelPipeline p = streamChannel.pipeline();
+    p.addLast(new Http2StreamFrameToHttpObjectCodec(false));
+    p.addLast(new HttpObjectAggregator(http2MultiplexedChannelPool.getMaxContentLength()));
+    p.addLast(new SimpleChannelInboundHandler<FullHttpResponse>() {
+      @Override
+      protected void channelRead0(ChannelHandlerContext ctx, FullHttpResponse msg) throws Exception {
+        ctx.channel().attr(RESPONSE_PROMISE).getAndSet(null).setSuccess(msg.content().retainedDuplicate());
+        // Stream channel can't be reused. Release it here.
+        ctx.channel()
+            .parent()
+            .attr(Http2MultiplexedChannelPool.HTTP2_MULTIPLEXED_CHANNEL_POOL)
+            .get()
+            .release(ctx.channel());
+      }
+    });
+    streamChannel.pipeline().addLast(new AmbrySendToHttp2Adaptor());
 
-    DefaultHttp2HeadersFrame headersFrame = new DefaultHttp2HeadersFrame(http2Headers, false);
-    DefaultHttp2DataFrame dataFrame = new DefaultHttp2DataFrame(byteBuf, true);
-    childChannel.write(headersFrame);
-    childChannel.write(dataFrame);
-    childChannel.flush();
+    responsePromise = streamChannel.eventLoop().newPromise();
+    streamChannel.attr(RESPONSE_PROMISE).set(responsePromise);
+    streamChannel.writeAndFlush(request).awaitUninterruptibly(3, TimeUnit.SECONDS);
```

Comment:
3 seconds seems a bit arbitrary, and `awaitUninterruptibly` returns a boolean indicating if the write succeeded or not. If the return value is false, we should throw out some exception so the caller would not proceeds to call `receive` method.